core/usig-ui.go

```diff
@@ -31,17 +31,10 @@ import (
 // Processing of UIs generated by the same replica is synchronized by
 // waiting to ensure each of the UIs is processed only once and in the
 // order of counter value. The return value new indicates if the UI
-// needs to be processed. The processing has to be completed by
-// corresponding uiReleaser invoked on the UI. It is safe to invoke
+// needs to be processed. If so, the processing has to be completed by
+// invoking the returned release function. It is safe to invoke
 // concurrently.
-type uiCapturer func(replicaID uint32, ui *usig.UI) (new bool)
-
-// uiReleaser synchronizes completion of sequential processing of USIG
-// unique identifier.
-//
-// It allows to continue sequential processing of UIs, synchronized by
-// uiCapturer. It is safe to invoke concurrently.
-type uiReleaser func(replicaID uint32, ui *usig.UI)
+type uiCapturer func(msg messages.MessageWithUI) (new bool, release func())
```

Comment:
Is `uiCapturer` going to be used for messages included in the view-change message? If so, then the comment of this interface should be updated for the following statement:
> Processing of UIs generated by the same replica 

Besides, why not to have a another function `parsedUiCapturer` to avoid parsing the UI message again?

Reply:
No, I don't think `uiCapturer` would be useful to process the messages embedded into the ViewChange message.

If we were to reuse the UI parsed during message validation, we would need some way to pass it from `messageValidator` to `messageProcessor`. I think a better strategy would be to make the messages in `messages` package smarter. I was thinking of defining go structures for the messages that would be populated with all the pieces unserialized, like parsed UI, while also storing the original serialized stuff to check signatures etc. That would allow to parse the UI once the message is converted from serialized from to the go structure, and then simply reuse it. It would even allow to make the core independent from a particular mechanism of serialization, which currently happens to be protobuf.